File: piwigo/Settings/Cells/SliderTableViewCell.m
//
//  SliderTableViewCell.m
//  piwigo
//
//  Created by Spencer Baker on 3/9/15.
//  Copyright (c) 2015 bakercrew. All rights reserved.
//

#import "SliderTableViewCell.h"

@interface SliderTableViewCell()

@property (nonatomic, strong) UILabel *sliderCount;

@end

@implementation SliderTableViewCell

-(instancetype)init
{
	self = [super init];
	if(self)
	{
		self.backgroundColor = [UIColor piwigoCellBackgroundColor];
		
		self.sliderName = [UILabel new];
		self.sliderName.translatesAutoresizingMaskIntoConstraints = NO;
		self.sliderName.font = [UIFont piwigoFontNormal];
		self.sliderName.textColor = [UIColor piwigoLeftLabelColor];
		self.sliderName.adjustsFontSizeToFitWidth = NO;
		[self.contentView addSubview:self.sliderName];
		[self.contentView addConstraint:[NSLayoutConstraint constraintCenterHorizontalView:self.sliderName]];
		
		self.slider = [UISlider new];
		self.slider.translatesAutoresizingMaskIntoConstraints = NO;
		self.slider.minimumValue = 10;
		self.slider.maximumValue = 200;
        self.slider.tintColor = [UIColor piwigoOrange];
        self.slider.thumbTintColor = [UIColor piwigoThumbColor];
		[self.slider addTarget:self action:@selector(sliderChanged) forControlEvents:UIControlEventValueChanged];
		[self.contentView addSubview:self.slider];
		[self.contentView addConstraint:[NSLayoutConstraint constraintCenterHorizontalView:self.slider]];
		
		self.sliderCount = [UILabel new];
		self.sliderCount.translatesAutoresizingMaskIntoConstraints = NO;
		self.sliderCount.textAlignment = NSTextAlignmentRight;
		self.sliderCount.font = [UIFont piwigoFontNormal];
		self.sliderCount.textColor = [UIColor piwigoRightLabelColor];
		self.sliderCount.adjustsFontSizeToFitWidth = NO;
		[self.contentView addSubview:self.sliderCount];
		[self.contentView addConstraint:[NSLayoutConstraint constraintCenterHorizontalView:self.sliderCount]];
		
		if (@available(iOS 11, *)) {
            [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-[label]-[slider]-[count]-|"
																				 options:kNilOptions
																				 metrics:nil
																				   views:@{@"label" : self.sliderName,
                                                                                           @"slider" : self.slider,
																						   @"count" : self.sliderCount
																						   }]];
        } else {
            [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-15-[label]-[slider]-[count]-15-|"
                                                                                     options:kNilOptions
                                                                                     metrics:nil
                                                                                       views:@{@"label" : self.sliderName,
                                                                                               @"slider" : self.slider,
                                                                                               @"count" : self.sliderCount
                                                                                               }]];
        }
	}
	return self;
}

-(void)sliderChanged
{
    NSInteger ref = (int)roundf(self.slider.minimumValue);
    NSInteger value = self.slider.value;
    NSInteger newValue = value - ((value - ref) % self.incrementSliderBy);
    self.slider.value = newValue;
	self.sliderCount.text = [NSString stringWithFormat:@"%@%@%@", self.sliderCountPrefix, @(newValue), self.sliderCountSuffix];
}

-(void)setSliderValue:(NSInteger)sliderValue
{
	self.slider.value = sliderValue;
	[self sliderChanged];
}

-(NSInteger)getCurrentSliderValue
{
	return self.slider.value;
}

@end

File: piwigo/Settings/Cells/SliderTableViewCell.swift
//
//  SliderTableViewCell.swift
//  piwigo
//
//  Created by Spencer Baker on 3/9/15.
//  Copyright (c) 2015 bakercrew. All rights reserved.
//
//  Converted to Swift 5 by Eddy Lelièvre-Berna on 12/04/2020.
//

import UIKit

typealias CellSliderBlock = (Float) -> Void

class SliderTableViewCell: UITableViewCell {

    @IBOutlet weak var sliderName: UILabel!
    @IBOutlet weak var slider: UISlider!
    @IBOutlet weak var sliderValue: UILabel!
    
    private var valuePrefix: String?
    private var valueSuffix: String?
    private var incrementSliderBy:Float = 0.0
    private var oldNberOfDays:Float = 0.0

    var cellSliderBlock: CellSliderBlock?

    func getCurrentSliderValue() -> Float {
        return slider?.value ?? 0
    }

    func configure(with title:String, value:Float, increment:Float,
                   minValue:Float, maxValue:Float, prefix:String, suffix:String) {

        // Background color and aspect
        backgroundColor = .piwigoColorCellBackground()

        // Slider title
        sliderName.font = .systemFont(ofSize: 17)
        sliderName.textColor = .piwigoColorLeftLabel()
        sliderName.text = title

        // Slider
        slider.value = value
        incrementSliderBy = increment
        slider.minimumValue = minValue
        slider.maximumValue = maxValue
        slider.tintColor = .piwigoColorOrange()
        slider.thumbTintColor = .piwigoColorThumb()

        // Slider value
        sliderValue.font = .systemFont(ofSize: 17)
        sliderValue.textColor = .piwigoColorRightLabel()
        valuePrefix = prefix
        valueSuffix = suffix
        updateDisplayedValue(value)
    }

    @IBAction func sliderChanged(_ sender: Any) {
        updateDisplayedValue(slider.value)
        (cellSliderBlock ?? {_ in return})(slider.value)
    }

    func updateDisplayedValue(_ value: Float) {
        if Int(incrementSliderBy) == AlbumVars.shared.recentPeriodKey {
            // Special case of recent period, value = index of kRecentPeriods
            // 1, 2, 3, … 19, 20, 25, 30, 40, 50, 60, 80, 99 days (same as Piwigo server)
            var indexOfPeriod:Int = Int(value.rounded(.toNearestOrAwayFromZero))
            indexOfPeriod = min(indexOfPeriod, AlbumVars.shared.recentPeriodList.count - 1)
            indexOfPeriod = max(0, indexOfPeriod)
            slider.value = Float(indexOfPeriod)
            sliderValue.text = String(format: (valueSuffix ?? ""), String(AlbumVars.shared.recentPeriodList[indexOfPeriod]))
        }
        else {
            let quotient = (value / incrementSliderBy).rounded(.towardZero) * incrementSliderBy
            let remainder = value.truncatingRemainder(dividingBy: incrementSliderBy)
            let normalisedReminder = (remainder / incrementSliderBy).rounded(.toNearestOrAwayFromZero)
            slider.value = quotient + normalisedReminder * incrementSliderBy
            sliderValue.text = "\(valuePrefix ?? "")\(NSNumber(value: slider.value))\(valueSuffix ?? "")"
        }
    }

    override func prepareForReuse() {
        super.prepareForReuse()
        sliderName.text = ""
        valuePrefix = ""
        valueSuffix = ""
    }
}
